Display of large scanned PDFs on IOS devices

CIE81 opened the pull request:
This fix is linked to the following issue: https://github.com/mozilla/pdf.js/issues/9474

For the RGB images, the idea is to resize the image by reducing each scale * scale box of pixels to one pixel. We are parsing the image with a square box of scale * scale and for each iteration we are calculating an average of the pixels color inside the box and create one pixel with the average value. The algorithm works perfect for images that have the height and width multiple of scale. To address images that do not fit in that category we have applied the algorithm separate to the main area of the image (the area that is divisible by scale) and to the side areas, the right side, the bottom side and the bottom right.
For the B&W images the idea is to resize the image by increasing the image depth. From a B& W image we are going to create a Gray-scale image. 
In a B&W image each bit represent a pixel. We have created an algorithm to scale by 2 and another one to scale by 4. 
To scale the image by 2 we've grouped each 2 consecutive bits from the current line and the next line (line + 1) in a square of 4 bits. Then we've created a new byte (pixel) by multiplying the number of bits set to 1 (white) with 256 (white value on a gray-scale image) and divided the result by 4 (because we have considered 4 pixels). 
To scale the image by 4 we've grouped each 4 consecutive bits from the current line with 4 consecutive bits from the next 3 lines below in a square of 16 bits. Then we've created a new byte (pixel) by multiplying the number of bits set to 1 with 256 and divided the result by 16 (because we've considered 16 pixels). 
The problem for the B&W scaling algorithms is when the image width is not divisible by 8 and/or image height is not divisible by scale. For those cases we've created a separate scaling algorithm which we will call for each side areas (right side, bottom side and bottom right). 
While an RGB image can be scaled by any factor, the B&W images can only be scaled by 2 or by 4. However a scale by 2 means the new image size will be 4 times smaller, and a scale by 4 will create a new image with a size of 16 times smaller. We chose to do the B&W scaling only by 2 and 4 because of performance.
Below are the PDFJS files that I've changed:
•	src/core/image.js
•	src/display/canvas.js
•	src/shared/util.js

Commit:
implement the display of large scanned PDFs on IOS

CIE81 commented:
Can someone please take a look at the Travis CI build fail ? I am not sure why is that failing.

Thank you,
Ioan

brendandahl commented:
The unit tests running in the node environment are failing. Looks like some of your code assumes it's always running in a browser.  run `gulp unittestcli` to reproduce

CIE81 commented:
I see the error but it doesn't tell me much to know where it's coming from

events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: spawn node_modules/.bin/jasmine ENOENT
    at exports._errnoException (util.js:1020:11)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:197:32)
    at onErrorNT (internal/child_process.js:376:16)
    at _combinedTickCallback (internal/process/next_tick.js:80:11)
    at process._tickCallback (internal/process/next_tick.js:104:9)

I've only changed 3 files, **image.js** and **util.js** don't have browser context, and in **canvas.js** I can't find any place in my changes where I would assume I run in a browser.

Can you please suggest anything that can give me more details on the error ?

Thank you,
Ioan

brendandahl commented:
https://travis-ci.org/mozilla/pdf.js/builds/364151853#L587

brendandahl commented:
Looks like you fixed the unitest, now just some lint errors. Were you able to run it locally? Seems to be working fine here.

CIE81 commented:
You were right about the browser assumption. I was using the code below to detect if we are running on a mobile device: return (/Mobi/.test(navigator.userAgent));
I have fixed it and will submit all the changes soon.

Thank you,
Ioan

CIE81 commented:
Hi, 

Did anyone got a chance to review the code ?

Thank you,
Ioan

yurydelendik commented:
Related solution is https://github.com/mozilla/pdf.js/pull/9255 . We don't want to use `navigator.userAgent` as a solution. Please derive scaling factor from canvas dpi.

yurydelendik reviewed (changes requested):
Make decision for image scaling based on the maximum resolution needed for UI to display the image.

yurydelendik commented on src/core/image.js:
```diff
@@ -196,6 +196,372 @@ var PDFImage = (function PDFImageClosure() {
+  // Determine if we run on an IOS Mobile device
+  PDFImage.isMobile = function() {
+    if (typeof navigator !== 'undefined') {
+      return (/Mobi/.test(navigator.userAgent));
```
> is there a way to calculate the image scale in image.js ?

yes, see #9255

CIE81 replied:
Sorry my question was not clear. What I wanted to say is how are we going to know if an image is going to render or not on a device, without trying to render it on a canvas on that device ? 
By calculating the image scale, I wanted to say, how do we automatically determine if the image needs to be scaled and by what factor.

CIE81 replied:
I can isolate the scaling functions in a new file scaling.js and call them in case it is needed from canvas.js. In order to determine if a document needs to be scaled I will try to render it, and if it does not render I will scale it by 2, then 4, 8, ... This way I would only apply the scaling in case I can't render the document. What do you think ?

Thank you,
Ioan

CIE81 commented:
Hi,

I have moved all the scaling code in a separate file and moved the entry point to canvas.js. Please take a look and let me know what you think.

The condition for scaling now is only if a canvas of the image size can't be rendered. This is device agnostic and will scale both image (B&W and RGB) and masks in case they can't be displayed by WebGL.

Thank you,
Ioan

Snuffleupagus commented:
Similar to e.g. PR #7047, it would seem to me that you'd really want PR #9255 to land *first*[1] and then base any further image optimizations on that functionality.

/cc @timvandermeij 

---

[1] Which is currently stuck on https://github.com/mozilla/pdf.js/pull/9255#issuecomment-423691024.

timvandermeij commented:
Closing since this is currently not in a mergeable state, it may cause image quality issues and it's not clear if it's even necessary if #9255 is finished. Let's await that and if this is still a problem we can always reopen/revisit. Thanks.

kinztechcom commented:
@timvandermeij 

> Closing since this is currently not in a mergeable state, it may cause image quality issues **and it's not clear if it's even necessary if #9255 is finished**. Let's await that and if this is still a problem we can always reopen/revisit. Thanks.

#9255 Only addresses image mask correct? This PR addresses more than just image mask does it not?

timvandermeij commented:
There is also code to scale masks, so that may not be necessary anymore after the PR above. Basically, after that PR a re-evaluation needs to be done for if there are still issues and if so, how they can be fixed.

kinztechcom commented:
@timvandermeij  It appears to definitely still be an issue for pdf's such as the first one provided in #9632 
After implementing both PR's locally the issue persisted with #9255 changes but was fixed with #9645 

I also feel as though #9255 does not provide a solution as elegant as this PR, but that's just my opinion. I suppose 9255 is more recently updated though and has a higher chance of going first. If 9255 goes through first, inline images will need an equivalent solution next.

timvandermeij commented:
Yes, if it turns out that the PR is not sufficient, then we can always revisit this. For that reason, we keep the original issue open.

CIE81 commented:
I've took the latest version of the pdfjs and the issue is not fix, neither for images, nor for masks.

Please revisit this, and let me know if you need me to merge the current version in this ticket.

Thank you,
Ioan

k1nger commented:
I would love for this to get merged, we've been trying to maintain a fork with this implementation trying to keep current.

kevinswartz commented:
Are these changes merged in somewhere?  I'd love a fix for this!  I see #10888 but I don't see much progress there.

kinztechcom commented:
@kevinswartz  They have not been. I have been just maintaining this PR's implementation on our private repo everytime a new pdfjs version is released. It's still strange to me that they chose #9255 over this PR.

kevinswartz commented:
Thanks @kinztechcom.  Bummer it's not merged yet.  I'm unable to display scanned pdfs on ios with the latest.

aarondunn commented:
@yurydelendik  can we re-evaluate the selected patch? The chosen one (#10967 ) does not work but this fix does when we apply it. We are using this for Musopen.org for tens of thousands of PDFs which are not viewable without this patch (#9645 ).  Thanks!

kinztechcom commented:
@timvandermeij Can we get another look at this pull request? I know at this point the implementation is most likely outdated and would need to be re-done. I've been manually adding this solution into current releases for almost 2 years. It's pretty annoying to have to consistently do that.
Or perhaps have I missed an accepted pull request and this issue has been **correctly** resolved elsewhere?
